src/libraries/System.Private.CoreLib/src/System/DateTime.cs

```diff
@@ -249,81 +315,507 @@ public DateTime(int year, int month, int day, int hour, int minute, int second,
+        /// </exception>
+        /// <remarks>
+        /// The allowable values for <paramref name="year"/>, <paramref name="month"/>, and <paramref name="day"/> parameters
+        /// depend on the <paramref name="calendar"/> parameter. An exception is thrown if the specified date and time cannot
+        /// be expressed using <paramref name="calendar"/>.
+        ///
+        /// For applications in which portability of date and time data or a limited degree of time zone awareness is important,
+        /// you can use the corresponding <see cref="DateTimeOffset"/> constructor.
+        /// </remarks>
+        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, int microsecond, Calendar calendar!!, DateTimeKind kind)
         {
             if ((uint)millisecond >= MillisPerSecond) ThrowMillisecondOutOfRange();
             if ((uint)kind > (uint)DateTimeKind.Local) ThrowInvalidKind();
 
             if (second != 60 || !s_systemSupportsLeapSeconds)
             {
-                ulong ticks = DateToTicks(year, month, day) + TimeToTicks(hour, minute, second);
-                ticks += (uint)millisecond * (uint)TicksPerMillisecond;
-                Debug.Assert(ticks <= MaxTicks, "Input parameters validated already");
+                var dateTime = calendar.ToDateTime(year, month, day, hour, minute, second, millisecond);
+                dateTime = dateTime.AddMicroseconds(microsecond);
+                ulong ticks = dateTime.UTicks;
                 _dateData = ticks | ((ulong)kind << KindShift);
             }
             else
             {
                 // if we have a leap second, then we adjust it to 59 so that DateTime will consider it the last in the specified minute.
-                this = new DateTime(year, month, day, hour, minute, 59, millisecond, kind);
+                this = new DateTime(year, month, day, hour, minute, 59, millisecond, microsecond, calendar, kind);
                 ValidateLeapSecond();
             }
         }
 
-        // Constructs a DateTime from a given year, month, day, hour,
-        // minute, and second for the specified calendar.
-        //
-        public DateTime(int year, int month, int day, int hour, int minute, int second, int millisecond, Calendar calendar!!)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static ulong Init(int year, int month, int day, int hour, int minute, int second, int millisecond)
         {
+            if ((uint)millisecond >= MillisPerSecond) ThrowMillisecondOutOfRange();
+
             if (second != 60 || !s_systemSupportsLeapSeconds)
             {
-                _dateData = calendar.ToDateTime(year, month, day, hour, minute, second, millisecond).UTicks;
+                ulong ticks = DateToTicks(year, month, day) + TimeToTicks(hour, minute, second);
+                ticks += (uint)millisecond * (uint)TicksPerMillisecond;
```

Comment:
>uint [](http://example.com/codeflow?start=26&length=4)

why we need the casting here?

Reply:
Because `millisecond` is `int` instead of `uint` & `TicksPerMillisecond` is `long` instead of `uint`

Reply:
We need this explicit casting because ticks is unsigned and this is not implicitly derived.
![grafik](https://user-images.githubusercontent.com/10373452/162796366-9e065b69-22eb-481d-88dc-77821b1f88e7.png)